src/client/Microsoft.Identity.Client/Internal/Requests/ClientCredentialRequest.cs

```diff
@@ -36,105 +37,144 @@ protected override async Task<AuthenticationResult> ExecuteAsync(CancellationTok
+                    if (proactivelyRefresh)
                     {
-                        cacheInfoTelemetry = CacheRefreshReason.NoCachedAccessToken;
+                        cacheInfoTelemetry = CacheRefreshReason.ProactivelyRefreshed;
+
+                        SilentRequestHelper.ProcessFetchInBackground(
+                        cachedAccessTokenItem,
+                        () => GetAccessTokenAsync(proactivelyRefresh, cancellationToken, logger), logger);
                     }
                 }
+                catch (MsalServiceException e)
+                {
+                    return await HandleTokenRefreshErrorAsync(e, cachedAccessTokenItem).ConfigureAwait(false);
+                }
             }
             else
             {
-                cacheInfoTelemetry = CacheRefreshReason.ForceRefreshOrClaims;
-                logger.Info("Skipped looking for an Access Token in the cache because ForceRefresh or Claims were set. ");
+                //  No AT in the cache 
+                if (AuthenticationRequestParameters.RequestContext.ApiEvent.CacheInfo != CacheRefreshReason.Expired)
+                {
+                    cacheInfoTelemetry = CacheRefreshReason.NoCachedAccessToken;
+                }
+                else
+                {
+                    cacheInfoTelemetry = CacheRefreshReason.Expired;
+                }
+
+                authResult = await GetAccessTokenAsync(false, cancellationToken, logger).ConfigureAwait(false);
+            }
+
+            AuthenticationRequestParameters.RequestContext.ApiEvent.CacheInfo = cacheInfoTelemetry;
+
+            return authResult;
+        }
+
+        private async Task<AuthenticationResult> GetAccessTokenAsync(
+            bool proactivelyRefresh, 
+            CancellationToken cancellationToken, 
+            ILoggerAdapter logger)
+        {
+            await ResolveAuthorityAsync().ConfigureAwait(false);
+            MsalTokenResponse msalTokenResponse;
+            AuthenticationResult authResult;
+            MsalAccessTokenCacheItem cachedAccessTokenItem = null;
+
+            //calls to AAD
+            if (ServiceBundle.Config.AppTokenProvider == null)
+            {
+                msalTokenResponse = await SendTokenRequestAsync(GetBodyParameters(), cancellationToken).ConfigureAwait(false);
+                return await CacheTokenResponseAndCreateAuthenticationResultAsync(msalTokenResponse).ConfigureAwait(false);
             }
 
-            if (AuthenticationRequestParameters.RequestContext.ApiEvent.CacheInfo == CacheRefreshReason.NotApplicable)
+            // Bypass cache and send request to token endpoint, when 
+            // 1. Force refresh is requested, or
+            // 2. Claims are passed, or 
+            // 3. If the AT needs to be refreshed pro-actively 
+            if (_clientParameters.ForceRefresh ||
+                proactivelyRefresh || 
+                !string.IsNullOrEmpty(AuthenticationRequestParameters.Claims))
```

Comment:
Don't we want to throttle these too since they can cause 429?

Reply:
@bgavrilMS should we?

Reply:
Yes, these will make HTTP request so they should be behind barrier.

I'm concerned about the background thread though. I wonder if the app can kill it or set it to low pri somehow which will leave the semaphore blocked. I wonder if we should set a timeout on that semaphore? 3-5 seconds...

Reply:
added `await s_semaphoreSlim.WaitAsync(cancellationToken).ConfigureAwait(false);` per our conversation